File: problems/LeetCode/backtrack/letterCombinations.py
MAPPING = ['','','abc','def','ghi','jkl','mno','pqrs','tuv','wxyz']
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        n = len(digits)

        if n == 0:
            return []

        ans = []
        path = [''] * n

        def backtrack(i: int):
            if i == n:
                ans.append(''.join(path))
                return 
            
            for c in MAPPING[int(digits[i])]:
                path[i] = c
                backtrack(i + 1)
        
        backtrack(0)
        return ans

File: problems/LeetCode/TopInterview150/letterCombinations.ts
function letterCombinations(digits: string): string[] {
  const mapping = [
    '',
    '',
    'abc',
    'def',
    'ghi',
    'jkl',
    'mno',
    'pqrs',
    'tuv',
    'wxyz',
  ];

  const ans: string[] = [];

  const backtrack = (idx: number, str: string) => {
    if (str.length === digits.length) {
      ans.push(str);
      return;
    }

    const digit = Number(digits[idx]);
    const comb = mapping[digit];

    for (const letter of comb) {
      backtrack(idx + 1, str + letter);
    }
  };

  backtrack(0, '');
  return ans;
}

// Time: O(n*4^n) or O(n*3^n)
// 4^n: combination, n: string concatenation
// Space: O(n)
